example.gpr

```diff
@@ -11,6 +11,23 @@ project Example is
       for Default_Switches ("Ada") use ("-Og", "-g", "-gnata");
    end Compiler;
 
+   package Recordflux is
+      for Output_Dir use "generated";
+   end Recordflux;
+
+   package Prove is
+      for Proof_Switches ("Ada") use (
+         "-j0",
+         "--prover=z3,cvc5,altergo,colibri",
```

Comment:
Why explicitly enable provers? Aren't they enabled by default?

Reply:
These are the recommended gnatprove settings from the RecordFlux User Guide (https://docs.adacore.com/live/wave/recordflux/html/recordflux_ug/20-overview.html#proving-and-running-the-result) that have been found to be most effective.

I originally ran gnatprove with the given project as is. The message about the problematic code in example.adb did come up pretty soon:

```
example.adb:35:19: medium: precondition might fail, cannot prove Valid_Predecessor (Ctx, Fld)
   35 |   Request_Message.Set_Kind (Context, Request_Store); 
```

However, many other locations in the automatically generated RecordFlux code got ‘medium’ level errors too. For example:

```
rflx-rflx_arithmetic.adb:32:37: medium: overflow check might fail …
rflx-rflx_arithmetic.adb:55:34: medium: assertion might fail …
rflx-rflx_arithmetic.adb:55:34: cannot prove Result <= 2 ** Bits - 2 ** Mask …
```

These were all due to unproven verification conditions. I then increased the number of parallel threads and gradually increased the number of steps quite a bit (eventually, I let it run overnight), but some contracts in the RecordFlux code were still unproven.

With these recommended settings, however, all the automatically generated code was fully proven and just the problematic manually written line example.adb:35 was reported. The process took about 18 minutes altogether on my machine.
 
I did not investigate which part had the biggest effect. Possibly, the leaving out of counterexamples. Here though the line tells the recommended order of provers to gnatprove, I believe. That probably makes some difference as well.

I used:

```
$ rflx --version
RecordFlux 0.10.0
attrs 21.4.0
icontract 2.6.2
pydantic 1.10.7
pydotplus 2.0.2
ruamel.yaml 0.17.26
z3-solver 4.12.2.0

$ gnatprove --version
SPARK Pro 24.0w (20221214)
```